sys/log/full/src/log_fcb2.c

```diff
@@ -0,0 +1,890 @@
+                }
+#endif
+            } else if (el_max_sec < el_min_sec) {
+                if (sec < el_max_sec || sec >= el_min_sec) {
+                    info->used += fcb->f_ranges[i].fsr_sector_size;
+                }
+#if MYNEWT_VAL(LOG_STORAGE_WATERMARK)
+                if (sec < el_max_sec || sec > fl->fl_watermark_off) {
+                    info->used_unread +=
+                      fcb->f_ranges[i].fsr_sector_size - fl->fl_watermark_off;
+                } else if (sec == fl->fl_watermark_sec) {
+                    info->used_unread +=
+                      fcb->f_ranges[i].fsr_sector_size - fl->fl_watermark_off;
+                }
+#endif
+            }
+            if (sec == el_max_sec) {
+                info->used += fcb->f_active.fe_data_off;
+#if MYNEWT_VAL(LOG_STORAGE_WATERMARK)
+                if (sec != fl->fl_watermark_sec) {
+                    info->used_unread += fcb->f_active.fe_data_off;
+                } else {
+                    info->used_unread +=
+                      fcb->f_active.fe_data_off - fl->fl_watermark_off;
+                }
+#endif
+            }
+        }
+    }
+    if (fl->fl_watermark_sec == FCB_SECTOR_OLDEST) {
+        info->used_unread = 0xffffffff;
+    }
+
+    os_mutex_release(&fcb->f_mtx);
+
+    return 0;
+}
+#endif
+
+#if MYNEWT_VAL(LOG_STORAGE_WATERMARK)
+static int
+log_fcb2_new_watermark_index(struct log *log, struct log_offset *log_off,
+                             void *dptr, uint16_t len)
+{
+    struct fcb_entry *loc;
+    struct fcb_log *fl;
+    struct log_entry_hdr ueh;
+    int rc;
+
+    loc = (struct fcb_entry *)dptr;
+    fl = (struct fcb_log *)log->l_arg;
+
+    rc = log_fcb2_read(log, loc, &ueh, 0, sizeof(ueh));
+
+    if (rc != sizeof(ueh)) {
+        return -1;
+    }
+
+    /* Set log watermark to end of this element */
+    if (ueh.ue_index >= log_off->lo_index) {
+        fl->fl_watermark_sec = loc->fe_sector;
+        fl->fl_watermark_off = loc->fe_data_off + loc->fe_data_len;
+        return -1;
+    } else {
+        return 0;
+    }
+}
+
+static int
+log_fcb2_set_watermark(struct log *log, uint32_t index)
+{
+    int rc;
+    struct log_offset log_offset;
+    struct fcb_log *fl;
+    struct fcb *fcb;
+
+    fl = (struct fcb_log *)log->l_arg;
+    fcb = &fl->fl_fcb;
+
+    log_offset.lo_arg = NULL;
+    log_offset.lo_ts = 0;
+    log_offset.lo_index = index;
+    log_offset.lo_data_len = 0;
+
+    /* Find where to start the walk, and set watermark accordingly */
+    rc = log_fcb2_walk(log, log_fcb2_new_watermark_index, &log_offset);
+    if (rc != 0) {
+        goto done;
+    }
+
+    /* If there are no entries to read and the watermark has not been set */
+    if (fl->fl_watermark_off == 0xffffffff) {
+        fl->fl_watermark_sec = fcb->f_oldest_sec;
+        fl->fl_watermark_off = 0;
+    }
+
+    return (0);
+done:
+    return (rc);
+}
+#endif
+
+/**
+ * Copies one log entry from source fcb to destination fcb
+ * @param src_fcb, dst_fcb
```

Comment:
quite unusual doxygen comment, usually @param refers to just one argument, other arguments are missing, same for the next function.